sync/syncer.go

```diff
@@ -198,89 +216,191 @@ func (s *Syncer) getLayerFromNeighbors(currenSyncLayer types.LayerID) (*types.La
 	return types.NewExistingLayer(types.LayerID(currenSyncLayer), blocksArr), nil
 }
 
-func (s *Syncer) fetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
-	output := s.fetchWithFactory(BlockReqFactory(blockIds), s.Concurrency)
+func (s *Syncer) FetchFullBlocks(blockIds []types.BlockID) ([]*types.Block, error) {
+	output := s.fetchWithFactory(NewBlockWorker(s, s.Concurrency, BlockReqFactory(), blockSliceToChan(blockIds)))
 	blocksArr := make([]*types.Block, 0, len(blockIds))
 	for out := range output {
 		block := out.(*types.Block)
-
-		eligible, err := s.BlockEligible(&block.BlockHeader)
-		if err != nil {
-			s.Warning(fmt.Sprintf("failed checking eligiblety %v", block.ID()), err)
-			continue
-		}
-		if !eligible {
-			s.Warning(fmt.Sprintf("block %v not eligible", block.ID()), err)
-			continue
-		}
-
-		if err := s.syncMissingContent(block); err != nil {
-			s.Error(fmt.Sprintf("failed derefrencing block data %v", block.ID()), err)
+		if err := s.SyncAndValidate(block); err != nil {
+			s.Error(fmt.Sprintf("failed derefrencing block data %v %v", block.ID(), err))
 			continue
 		}
 
 		s.Info("added block to layer %v", block.Layer())
 		blocksArr = append(blocksArr, block)
-
 	}
 
 	return blocksArr, nil
 }
 
-func (s *Syncer) syncMissingContent(blk *types.Block) error {
+func (s *Syncer) SyncAndValidate(blk *types.Block) error {
+	s.Info("sync and validate block %v ", blk.ID())
+
+	//check if known
+	if _, err := s.GetBlock(blk.Id); err == nil {
+		s.Info("we already know this block %v ", blk.ID())
+		return nil
+	}
+
+	//check block signature and is identity active
+	if eligable, err := s.IsIdentityActive(blk.MinerID.Key, blk.Layer()); err != nil || !eligable {
```

Comment:
the name `eligible` is wrong here, you're checking if the miner is active, not eligible

Reply:
Also, the miner's ID that is passed here should be extracted from the block's signature and not from the ID field (which should be removed ASAP)

Reply:
this name comes from another pkg i will be happy to change it
do you have a more appropriate name ?